Objects/regions.c

```diff
@@ -92,9 +114,265 @@ struct regionmetadata {
+    Py_region_ptr_t parent = Py_region_ptr(self->parent);
+    if (parent != 0 && !IS_CONST_REGION(parent)) {
+        regionmetadata_dec_rc(parent);
+    }
+
+merge:
+    if (!IS_CONST_REGION(other)) {
+        regionmetadata_inc_rc(other);
+    }
+
+    self->parent = Py_region_ptr_with_tags(other | Py_METADATA_MERGE_TAG);
+    // No decref, since this is a week reference. Otherwise we would get
+    // a cycle on a non GC'ed object and the bridge.
+    self->bridge = NULL;
+    Py_RETURN_NONE;
+
+fail:
+    throw_region_error(self->bridge, _Py_CAST(regionmetadata*, other)->bridge,
+                       "Unable to merge regions with differen parents", NULL);
+    return NULL;
+}
+#define regionmetadata_merge(self, other) \
+  (regionmetadata_merge(self, _Py_CAST(Py_region_ptr_t, other)));
+
+static bool is_bridge_object(PyObject *op) {
+    regionmetadata* region = regionmetadata_get_representative_or_null(Py_REGION(op));
+    // Constant regions (indicated by NULL) never have a bridge object.
+    if (!region) {
+        return false;
+    }
+
+    // It's not yet clear how immutability will interact with region objects.
+    // It's likely that the object will remain in the object topology but
+    // will use the properties of a bridge object. This therefore checks if
+    // the object is equal to the regions bridge object rather than checking
+    // that the type is `PyRegionObject`
+    return _PyObject_CAST(region->bridge) == op;
+}
+
+static void regionmetadata_open(regionmetadata* self) {
+    self = regionmetadata_get_representative_or_null(self);
+    if (self) {
+        self->is_open = 1;
+    }
+}
+
+static bool regionmetadata_is_open(regionmetadata* self) {
+    self = regionmetadata_get_representative_or_null(self);
+    if (!self) {
+        // All constant regions are open
+        return true;
+    }
+
+    return self->is_open == 0;
+}
+
+static bool regionmetadata_has_ancestor(regionmetadata* self, regionmetadata* other) {
+    other = regionmetadata_get_representative_or_null(other);
+
+    // Constant regions can't be an ancestor
+    if (!other) {
+        return false;
+    }
+
+    // Make sure we actually look at our representative
+    self = regionmetadata_get_representative_or_null(self);
+    while (self) {
+        if (self == other) {
+            return true;
+        }
+        self = regionmetadata_get_parent(self);
+    }
+    return false;
+}
+
+int _Py_IsLocal(PyObject *op) {
+    if (IS_DEFAULT_REGION(Py_REGION(op))) {
+        return true;
+    }
+
+    if (IS_IMMUTABLE_REGION(Py_REGION(op))) {
+        return false;
+    }
+
+    Py_region_ptr_t region = regionmetadata_get_representative(Py_REGION(op));
+
+    // Here we set the region to the representative we received.
+    // In most cases this will do nothing, but in others it can decrese
+    // future lookup times.
+    Py_SET_REGION(op, region);
+    if (IS_DEFAULT_REGION(region)) {
+        return true;
+    }
```

Comment:
This will need to update reference counts.

Reply:
The ref increase is done by `Py_SET_REGION`.